### PR TITLE
refactor: add some base classes to refactor pegasus_value_schema

### PR DESCRIPTION
**Pull request by levy5307**
### What problem does this PR solve? 
Add some base classes to refactor pegasus_value_schema, for example: `pegasus_value_schema`, `pegasus_value_segment` and `pegasus_value_manager`.

The global design is here:
image

### Check List 

Tests 

- Unit test
No node actually. I will add unit tests in later pull requests.

### CODE REVIEW

File: src/base/pegasus_value_segment.h [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#pragma once
+
+namespace pegasus {
+
+enum segment_type
```
Comment (neverchanje) [inline src/base/pegasus_value_segment.h:24]:
```suggestion
enum value_field_type
```

I would recommend calling this "field". Actually, no system that I have seen so far is calling a "field" to "segment".

File: src/base/pegasus_value_segment.h [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+{
+    /// TBD(zlw)
+};
+
+struct pegasus_value_segment
+{
+    virtual ~pegasus_value_segment() = default;
+    virtual segment_type type() = 0;
+};
```
Comment (neverchanje) [inline src/base/pegasus_value_segment.h:33]:
```suggestion
struct value_field
{
    virtual ~value_field() = default;
    virtual value_field_type type() = 0;
};
```

File: src/base/pegasus_value_schema.h [hunk 3]
```diff
@@ -234,4 +236,31 @@ class pegasus_value_generator
...
+{
+    /// TBD(zlw)
+};
+
+struct generate_value_params
+{
+    /// TBD(zlw)
+};
+
+class pegasus_value_schema
```
Comment (neverchanje) [inline src/base/pegasus_value_schema.h:249]:
```suggestion
class value_schema
```

File: src/base/value_manager.h [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+#include "pegasus_value_schema.h"
+#include <dsn/utility/singleton.h>
+
+namespace pegasus {
+
+class value_manager : public dsn::utils::singleton<value_manager>
+{
+public:
+    void register_schema(value_schema *schema);
+    /// using raw value in rocksdb and data version stored in meta column family to get data version
```
Comment (Smityz) [inline src/base/value_manager.h:31]:
```suggestion
    /// using the raw value in rocksdb and data version stored in meta column family to get data version
```

File: src/base/value_manager.h [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+ */
+
+#pragma once
+
+#include "pegasus_value_schema.h"
+#include <dsn/utility/singleton.h>
+
+namespace pegasus {
+
+class value_manager : public dsn::utils::singleton<value_manager>
```
Comment (neverchanje) [inline src/base/value_manager.h:27]:
```suggestion
class value_schema_manager : public dsn::utils::singleton<value_manager>
```

value_schema_manager is more accurate

File: src/base/value_schema_manager.h [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    /// version
+    value_schema *get_value_schema(uint32_t meta_cf_data_version, dsn::string_view value) const;
+    value_schema *get_value_schema(uint32_t version) const;
+    value_schema *get_latest_value_schema() const;
+
+private:
+    value_schema_manager() = default;
+    friend class dsn::utils::singleton<value_schema_manager>;
+
+    std::map<data_version, value_schema *> _schemas;
```
Comment (neverchanje) [inline src/base/value_schema_manager.h:41]: Use array instead of map. You need to ensure the cost is as least as possible.

Comment (levy5307) [reply src/base/value_schema_manager.h:41]: There won't be too many schemas, so I think the cost won't be too much.
And I use std::map here is to maintain the order for these schemas,  which will be useful in function `get_latest_value_schema`

**Review by neverchanje — APPROVED** [review]

**Review by Shuo-Jia — APPROVED** [review]

**Pull request merged**